### DISCUSSION TITLE
`JVMTI` for Wasm

### DISCUSSION THREAD

**Discussion by ejrgilbert** [dylibso/wasi-observe#19] [Ideas]
Here's a brief summary of what I was discussing at the meeting today. Basically, I'm looking into what it would mean to have a [`JVMTI`](https://docs.oracle.com/javase/8/docs/platform/jvmti/jvmti.html#whatIs) for WebAssembly. So, the engine would directly support instrumenting a module dynamically through monitors interfacing with an engine and attaching callbacks to certain observable "events" as they happen at runtime.

The engine would have a notion of an application and its _subscribing_ Wasm modules (the monitors). These monitors would use the ABI to tell the engine which events its interested in and which callbacks to call when the events occur dynamically. For events observable in bytecode (function enter/exit, the `call` opcode, etc.), the monitor could _reflect_ on the application to find the points its interested in and attach a callback to that point in the bytecode. For events _not observable in bytecode_, like "application start", "application exit", "garbage collect start", "thread start", etc. there would be some ABI calls that would directly attach callbacks to those events occurring (not tied to the bytecode)

**For example:** let's say I want to create a flame graph of my module based on some input. My monitor would interface with the engine to walk the application bytecode, attaching callbacks to function entry/exit that keep track of function visitation history. Then, it would attach a callback to "application exit" that flushes the history in some format that can be rendered as a flame graph visualization.

**For some motivation:** The monitor is application and source language agnostic, enabling it to be reused across engines that provide the ABI. This injection strategy also keeps from having to instrument the bytecode offline (through bytecode rewriting). It can be injected _as the application runs_, so if an engine were to design this well, a system maintainer could turn on/off monitoring in production to help diagnose an issue _as its happening in production_.

**Comment by calebschoepp** [comment]
Thanks @ejrgilbert. Apologies for the basic questions, but I'm very new to this dynamic monitoring concept and want to really make sure I'm understanding what you're explaining.

- Is it safe for me to replace engine with Wasm runtime in my head or do you mean something else?
- I'm a bit confused about what the monitors are. At one point you say "...subscribing Wasm modules (the monitors)". Does this mean the monitors are literally just the Wasm modules provided by the end user? Are the monitors some kind of wrapper around the module? Or something entirely else that I'm not understanding.
- Is your plan to have this work for components too? Or just modules?

**Reply by ejrgilbert** [reply]
Hi! Sorry, I forgot to respond yesterday.

1. Yes, I tend to use them interchangeably in this context.
2. Right! The monitors are stand-alone Wasm modules that can instrument an arbitrary application (Wasm module or component). The monitor modules interface with the engine _in some way_ to instruct the engine on how to attach callbacks to the dynamic execution of the application. This can be at the bytecode instruction level (`br`, `call`, etc.) _OR_ higher level granularities (function entry/exit) _OR_ engine-observable operations (thread ops, GC ops, etc.). This is different than instrumenting applications through directly injecting logic since the engine itself manages your instrumentation. This approach removes the need to preprocess applications you're wanting to monitor, enabling the dynamic instrumentation of applications. It also enables the engine to do some optimizations through intrinsically handing instrumented application events.
3. Yes, we're hoping to support both components and modules. This is still very much in the works though. I was mainly looking at `wasi-observe` to see if this would be a good place for defining the engine API that would enable the monitoring of applications through interfacing with an engine dynamically. After our meeting, maybe it should be in a different place since you're all discussing source-level instrumentation though.

Hopefully this answers your questions, we can also meet again over Zoom if you think questions would be answered better conversationally.

**Reply by calebschoepp** [reply]
Thanks @ejrgilbert that clear some things up for me.

> After our meeting, maybe it should be in a different place since you're all discussing source-level instrumentation though.
I think that makes sense to me. To be honest I'm not sure where the right place for defining the engine API would be. Is there any way I can help you find a home for this? I want to make sure you don't just get sent away b/c WASI Observe is "not the right place" for what you're working on.

**Reply by ejrgilbert** [reply]
@calebschoepp Glad it cleared things up! I might just bring this up during a CG meeting to figure out how we can move this effort forward. Thanks for discussing and working with me!
